File: frontend/src/lib/api/my.ts
import { apiClient, type ApiResult, type PaginatedData } from './client';
import { get } from 'svelte/store';
import { t } from 'svelte-i18n';

export type MyIssueStatus = 'backlog' | 'todo' | 'in_progress' | 'done';
export type MyIssuePriority = 'low' | 'medium' | 'high' | 'urgent';

export interface MyIssue {
	id: string;
	project_id: string;
	workspace_id: string;
	project_name: string;
	title: string;
	status: MyIssueStatus;
	priority: MyIssuePriority;
	assignee_id?: string;
	updated_at: string;
}

export interface MyActivity {
	id: string;
	issue_id: string;
	project_id: string;
	workspace_id: string;
	project_name: string;
	issue_title: string;
	user_id?: string;
	author_name?: string;
	action: string;
	detail: Record<string, unknown>;
	created_at: string;
}

function mapIssue(raw: Record<string, unknown>): MyIssue {
	return {
		id: String(raw.id ?? ''),
		project_id: String(raw.project_id ?? ''),
		workspace_id: String(raw.workspace_id ?? ''),
		project_name: String(raw.project_name ?? ''),
		title: String(raw.title ?? ''),
		status: String(raw.state ?? raw.status ?? 'todo') as MyIssueStatus,
		priority: String(raw.priority ?? 'medium') as MyIssuePriority,
		assignee_id: raw.assignee_id ? String(raw.assignee_id) : undefined,
		updated_at: String(raw.updated_at ?? '')
	};
}

function mapActivity(raw: Record<string, unknown>): MyActivity {
	const detailRaw = raw.detail ?? raw.details ?? raw.payload;
	const detail = detailRaw && typeof detailRaw === 'object' ? (detailRaw as Record<string, unknown>) : {};
	return {
		id: String(raw.id ?? ''),
		issue_id: String(raw.issue_id ?? raw.resource_id ?? ''),
		project_id: String(raw.project_id ?? ''),
		workspace_id: String(raw.workspace_id ?? ''),
		project_name: String(raw.project_name ?? ''),
		issue_title: String(raw.issue_title ?? ''),
		user_id: raw.user_id ? String(raw.user_id) : raw.actor_id ? String(raw.actor_id) : undefined,
		author_name: raw.author_name ? String(raw.author_name) : undefined,
		action: String(raw.action ?? raw.event_type ?? ''),
		detail,
		created_at: String(raw.created_at ?? '')
	};
}

export const myApi = {
	async getMyIssues(params?: { page?: number; per_page?: number }): Promise<ApiResult<PaginatedData<MyIssue>>> {
		const queryParams = new URLSearchParams();
		if (params?.page) queryParams.set('page', String(params.page));
		if (params?.per_page) queryParams.set('per_page', String(params.per_page));
		const queryString = queryParams.toString();
		const endpoint = queryString ? `/api/v1/my/issues?${queryString}` : '/api/v1/my/issues';
		
		const result = await apiClient.get<PaginatedData<Record<string, unknown>>>(endpoint);
		if (result.code !== 0 || !result.data) {
			return { code: result.code, message: result.message || get(t)('api.loadMyIssuesFailed'), data: null };
		}
		return {
			code: 0,
			message: result.message,
			data: {
				...result.data,
				items: Array.isArray(result.data.items) ? result.data.items.map(mapIssue) : []
			}
		};
	},

	async getMyActivities(params?: { page?: number; per_page?: number }): Promise<ApiResult<PaginatedData<MyActivity>>> {
		const queryParams = new URLSearchParams();
		if (params?.page) queryParams.set('page', String(params.page));
		if (params?.per_page) queryParams.set('per_page', String(params.per_page));
		const queryString = queryParams.toString();
		const endpoint = queryString ? `/api/v1/my/activities?${queryString}` : '/api/v1/my/activities';
		
		const result = await apiClient.get<PaginatedData<Record<string, unknown>>>(endpoint);
		if (result.code !== 0 || !result.data) {
			return { code: result.code, message: result.message || get(t)('api.loadRecentActivityFailed'), data: null };
		}
		return {
			code: 0,
			message: result.message,
			data: {
				...result.data,
				items: Array.isArray(result.data.items) ? result.data.items.map(mapActivity) : []
			}
		};
	}
};

File: apps/api/src/routes/my.rs
use axum::{
    Extension,
    extract::{Query, State},
    response::IntoResponse,
};
use platform::{app::AppState, auth::JwtClaims};
use sea_orm::{DbBackend, FromQueryResult, Statement};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{
    error::ApiError,
    response::{ApiResponse, PaginatedData},
};

#[derive(Debug, Deserialize)]
pub struct PaginationQuery {
    #[serde(default = "default_page")]
    pub page: i64,
    #[serde(default = "default_per_page")]
    pub per_page: i64,
}

fn default_page() -> i64 {
    1
}

fn default_per_page() -> i64 {
    10
}

#[derive(Debug, Serialize)]
pub struct MyIssueResponse {
    pub id: Uuid,
    pub project_id: Uuid,
    pub workspace_id: Uuid,
    pub project_name: String,
    pub title: String,
    pub state: String,
    pub priority: String,
    pub assignee_id: Option<Uuid>,
    pub updated_at: String,
}

#[derive(Debug, Serialize)]
pub struct MyActivityResponse {
    pub id: Uuid,
    pub issue_id: Uuid,
    pub project_id: Uuid,
    pub workspace_id: Uuid,
    pub project_name: String,
    pub issue_title: String,
    pub user_id: Option<Uuid>,
    pub author_name: Option<String>,
    pub action: String,
    pub detail: serde_json::Value,
    pub created_at: String,
}

/// GET /api/v1/my/issues - List issues assigned to current user across workspaces
pub async fn get_my_issues(
    State(state): State<AppState>,
    Extension(claims): Extension<JwtClaims>,
    Query(pagination): Query<PaginationQuery>,
) -> Result<impl IntoResponse, ApiError> {
    let user_id = Uuid::parse_str(&claims.sub)
        .map_err(|_| ApiError::Unauthorized("invalid user id".to_string()))?;

    let page = pagination.page.max(1);
    let per_page = pagination.per_page.clamp(1, 100);
    let offset = (page - 1) * per_page;

    #[derive(Debug, FromQueryResult)]
    struct MyIssueRow {
        id: Uuid,
        project_id: Uuid,
        workspace_id: Uuid,
        project_name: String,
        title: String,
        state: String,
        priority: String,
        assignee_id: Option<Uuid>,
        updated_at: chrono::DateTime<chrono::Utc>,
    }

    #[derive(Debug, FromQueryResult)]
    struct CountRow {
        count: i64,
    }

    let total_result = CountRow::find_by_statement(Statement::from_sql_and_values(
        DbBackend::Postgres,
        r#"
            SELECT COUNT(*) as count
            FROM work_items wi
            INNER JOIN projects p ON wi.project_id = p.id
            WHERE wi.assignee_id = $1
              AND (
                EXISTS (
                    SELECT 1
                    FROM workspace_members wm
                    WHERE wm.workspace_id = p.workspace_id
                      AND wm.user_id = $1
                )
                OR EXISTS (
                    SELECT 1
                    FROM workspace_bots wb
                    WHERE wb.workspace_id = p.workspace_id
                      AND wb.id = $1
                )
              )
        "#,
        vec![user_id.into()],
    ))
    .one(&state.db)
    .await?;

    let total = total_result.map(|r| r.count).unwrap_or(0);

    let issues = MyIssueRow::find_by_statement(Statement::from_sql_and_values(
        DbBackend::Postgres,
        r#"
            SELECT wi.id,
                   wi.project_id,
                   p.workspace_id,
                   p.name AS project_name,
                   wi.title,
                   wi.state,
                   wi.priority,
                   wi.assignee_id,
                   wi.updated_at
            FROM work_items wi
            INNER JOIN projects p ON wi.project_id = p.id
            WHERE wi.assignee_id = $1
              AND (
                EXISTS (
                    SELECT 1
                    FROM workspace_members wm
                    WHERE wm.workspace_id = p.workspace_id
                      AND wm.user_id = $1
                )
                OR EXISTS (
                    SELECT 1
                    FROM workspace_bots wb
                    WHERE wb.workspace_id = p.workspace_id
                      AND wb.id = $1
                )
              )
            ORDER BY wi.updated_at DESC
            LIMIT $2 OFFSET $3
        "#,
        vec![
            user_id.into(),
            (per_page as i64).into(),
            (offset as i64).into(),
        ],
    ))
    .all(&state.db)
    .await?;

    let items: Vec<MyIssueResponse> = issues
        .into_iter()
        .map(|issue| MyIssueResponse {
            id: issue.id,
            project_id: issue.project_id,
            workspace_id: issue.workspace_id,
            project_name: issue.project_name,
            title: issue.title,
            state: issue.state,
            priority: issue.priority,
            assignee_id: issue.assignee_id,
            updated_at: issue.updated_at.to_rfc3339(),
        })
        .collect();

    let total_pages = if total == 0 {
        0
    } else {
        (total + per_page - 1) / per_page
    };

    Ok(ApiResponse::success(PaginatedData {
        items,
        total,
        page,
        per_page,
        total_pages,
    }))
}

/// GET /api/v1/my/activities - List recent workspace activities visible to current user
pub async fn get_my_activities(
    State(state): State<AppState>,
    Extension(claims): Extension<JwtClaims>,
    Query(pagination): Query<PaginationQuery>,
) -> Result<impl IntoResponse, ApiError> {
    let user_id = Uuid::parse_str(&claims.sub)
        .map_err(|_| ApiError::Unauthorized("invalid user id".to_string()))?;

    let page = pagination.page.max(1);
    let per_page = pagination.per_page.clamp(1, 100);
    let offset = (page - 1) * per_page;

    #[derive(Debug, FromQueryResult)]
    struct MyActivityRow {
        id: Uuid,
        issue_id: Uuid,
        project_id: Uuid,
        workspace_id: Uuid,
        project_name: String,
        issue_title: String,
        user_id: Option<Uuid>,
        author_name: Option<String>,
        action: String,
        detail: serde_json::Value,
        created_at: chrono::DateTime<chrono::Utc>,
    }

    #[derive(Debug, FromQueryResult)]
    struct CountRow {
        count: i64,
    }

    let total_result = CountRow::find_by_statement(Statement::from_sql_and_values(
        DbBackend::Postgres,
        r#"
            SELECT COUNT(*) as count
            FROM activities a
            INNER JOIN work_items wi ON wi.id = COALESCE(a.issue_id, CASE WHEN a.resource_type = 'issue' THEN a.resource_id END)
            INNER JOIN projects p ON wi.project_id = p.id
            WHERE (
                EXISTS (
                    SELECT 1
                    FROM workspace_members wm
                    WHERE wm.workspace_id = p.workspace_id
                      AND wm.user_id = $1
                )
                OR EXISTS (
                    SELECT 1
                    FROM workspace_bots wb
                    WHERE wb.workspace_id = p.workspace_id
                      AND wb.id = $1
                )
            )
        "#,
        vec![user_id.into()],
    ))
    .one(&state.db)
    .await?;

    let total = total_result.map(|r| r.count).unwrap_or(0);

    let activities = MyActivityRow::find_by_statement(Statement::from_sql_and_values(
        DbBackend::Postgres,
        r#"
            SELECT a.id,
                   COALESCE(a.issue_id, CASE WHEN a.resource_type = 'issue' THEN a.resource_id END) AS issue_id,
                   wi.project_id,
                   p.workspace_id,
                   p.name AS project_name,
                   wi.title AS issue_title,
                   COALESCE(a.user_id, a.actor_id) AS user_id,
                   u.name AS author_name,
                   COALESCE(a.action, a.event_type) AS action,
                   COALESCE(a.detail, a.payload, '{}'::jsonb) AS detail,
                   a.created_at
            FROM activities a
            INNER JOIN work_items wi ON wi.id = COALESCE(a.issue_id, CASE WHEN a.resource_type = 'issue' THEN a.resource_id END)
            INNER JOIN projects p ON wi.project_id = p.id
            LEFT JOIN users u ON u.id = COALESCE(a.user_id, a.actor_id)
            WHERE (
                EXISTS (
                    SELECT 1
                    FROM workspace_members wm
                    WHERE wm.workspace_id = p.workspace_id
                      AND wm.user_id = $1
                )
                OR EXISTS (
                    SELECT 1
                    FROM workspace_bots wb
                    WHERE wb.workspace_id = p.workspace_id
                      AND wb.id = $1
                )
            )
            ORDER BY a.created_at DESC
            LIMIT $2 OFFSET $3
        "#,
        vec![user_id.into(), (per_page as i64).into(), (offset as i64).into()],
    ))
    .all(&state.db)
    .await?;

    let items: Vec<MyActivityResponse> = activities
        .into_iter()
        .map(|activity| MyActivityResponse {
            id: activity.id,
            issue_id: activity.issue_id,
            project_id: activity.project_id,
            workspace_id: activity.workspace_id,
            project_name: activity.project_name,
            issue_title: activity.issue_title,
            user_id: activity.user_id,
            author_name: activity.author_name,
            action: activity.action,
            detail: activity.detail,
            created_at: activity.created_at.to_rfc3339(),
        })
        .collect();

    let total_pages = if total == 0 {
        0
    } else {
        (total + per_page - 1) / per_page
    };

    Ok(ApiResponse::success(PaginatedData {
        items,
        total,
        page,
        per_page,
        total_pages,
    }))
}
